assets/js/blocks/stock-filter/block.js

```diff
@@ -0,0 +1,285 @@
+/**
+ * External dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import { speak } from '@wordpress/a11y';
+import { usePrevious, useShallowEqual } from '@woocommerce/base-hooks';
+import {
+	useQueryStateByKey,
+	useQueryStateByContext,
+	useCollectionData,
+} from '@woocommerce/base-context/hooks';
+import { getSetting } from '@woocommerce/settings';
+import { useCallback, useEffect, useState, useMemo } from '@wordpress/element';
+import CheckboxList from '@woocommerce/base-components/checkbox-list';
+import FilterSubmitButton from '@woocommerce/base-components/filter-submit-button';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+import { decodeEntities } from '@wordpress/html-entities';
+
+/**
+ * Internal dependencies
+ */
+import Label from './label';
+import { previewOptions } from './preview';
+import './style.scss';
+
+const STOCK_STATUS_OPTIONS = getSetting( 'stockStatusOptions', [] );
+const hideOutOfStockItems = getSetting( 'hideOutOfStockItems', false );
+if ( hideOutOfStockItems ) {
+	delete STOCK_STATUS_OPTIONS.outofstock;
+}
+const initialOptions = Object.entries(
+	STOCK_STATUS_OPTIONS
+).map( ( [ slug, name ] ) => ( { slug, name } ) );
```

Comment:
One problem with doing this is it mutates the `stockStatusOptions` object. So anywhere in the code that `getSetting( 'stockStatusOptions' )` is called will have the mutated object. We don't force immutability of values returned from `getSetting` (something we might want to consider doing in future iterations). 

A better option here might be to to do something like this to remove the `outofstock` property via something like:

```suggestion
const hideOutOfStockItems = getSetting( 'hideOutOfStockItems', false );
const { outofstock, ...otherStockStatusOptions } = getSetting( 'stockStatusOptions', {} );
let STOCK_STATUS_OPTIONS = hideOutOfStockItems ? otherStockStatusOptions : { outofstock, ...otherStockStatusOptions };
// filter added to handle if there are slugs without a corresponding name defined.
const initialOptions = Object.entries(
	STOCK_STATUS_OPTIONS
).map( ( [ slug, name ] ) => ( { slug, name } ) ).filter( ( status ) => !! status.name );
```

Reply:
Yes it is my bad, this case fixed as you suggested. Just changed `let STOCK_STATUS_OPTIONS` to `const` and sorted initial options alphabetically fro better UI performance.